### ISSUE TITLE
Resume policy doesn't resume

### ISSUE THREAD

**Issue by laurentiubanica** [janmg/logstash-input-azure_blob_storage#22]
Hi,
With the following configuration set:
input {
    azure_blob_storage {
        connection_string => ""        
	container => "insights-logs-networksecuritygroupflowevent"
        #registry_create_policy => "resume"
        logtype => "nsgflowlog"
        prefix => "resourceId=/"
        interval => 60
    }
}
I have noticed that the plugin goes into an endless loop -> After it reaches the end of registry.dat, it lists again all the blobs in the storage accounts and starts writing again the older events that have been previously already processed. -> this affects the way the events are reaching, after a while not being able to see the ones for the current hour, for example.
Somehow, the resume policy doesn't resume, but goes back to the beginning.

**Comment by janmg** [comment]
Can you provide more information on where the the registry is lost? Logstash starts "run(queue)" which does 2 things. Initialize based on the configs. For resume it loads the registry file into a Hash @registry. This hash contains a filename, the offset in bytes that have been already processed and the file length of the last time the file was listed.

https://github.com/janmg/logstash-input-azure_blob_storage/blob/master/lib/logstash/inputs/azure_blob_storage.rb#L124

Next, by design the process runs in an endless while loop, the plugin lists the files, does a lookup in the registry to fill the offsets  and creates a temporary newreg, which is then promoted to the new registry. This was changed recently so that historical files that do not exist anymore are no longer tracked in the registry.

As long as the plugin did not get a stop signal from logstash, the new registry is than selected on where the offset is smaller then the file length. So if the file has grown or the file has been created it will be put on the worklist. For new files a full_read is done and for grown files a partial_read of the new bytes are done.

If the processing was faster than the interval, the registry is saved in a separate thread and the remaining time for the interval is spent sleeping before listing the files again.

https://github.com/janmg/logstash-input-azure_blob_storage/blob/master/lib/logstash/inputs/azure_blob_storage.rb#L194

Log entries are printed based on what the code is doing at what time. There is a configuration for creating for the first x amount debug log entries, which will produce more output and will switch back to info after some time.

I need more information on where the registry is loosing track of the files. You can run a second logstash instance with different configuration, especially different registry file and write to ruby output to get a much clearer understanding of what is happening.

**Comment by laurentiubanica** [comment]
Actions performed:
1. output to stdout with rubydebug
2. the blob contained at the beginning info for h=18 -> I left LS running several cycles of 60 seconds with this blob content, then
3. I added to the blob a new hour h=19 -> I left LS running several cycles. A new line has been added to the registry.dat, then
4. I added to the blob a new hour h=20 and also left LS running several cycles of 60 s. A new line has been added to the registry.dat.

**Comment by laurentiubanica** [comment]
Attaching the logs and registry file.
[registry.txt](https://github.com/janmg/logstash-input-azure_blob_storage/files/7809116/registry.txt)
[log.zip](https://github.com/janmg/logstash-input-azure_blob_storage/files/7809117/log.zip)

**Comment by laurentiubanica** [comment]
What I noticed from the log is that, if no new data was appended the the blob, the plugin started again to write to stdout the older events, even though the message was : 
[INFO ] 2022-01-04 16:26:35.500 [[main]<azure_blob_storage] azureblobstorage - going to sleep for 60 seconds

**Comment by laurentiubanica** [comment]
I attached a new log, resulted by running LS for couple of minutes with registry containing all 3 offsets.
[newlog.zip](https://github.com/janmg/logstash-input-azure_blob_storage/files/7809250/newlog.zip)

**Comment by laurentiubanica** [comment]
I've performed a new test.
I have kept the same configuration, but decreased the interval, from 60 s to 1s.
I started logstash.
After few moments, I added new blobs to the storage, then I waited a while.
Logstash was logging the following:

[INFO ] 2022-01-05 14:29:51.505 [[main]<azure_blob_storage] azureblobstorage - worklist contains 5 blobs
[INFO ] 2022-01-05 14:29:51.604 [[main]<azure_blob_storage] azureblobstorage - Skipped writing the registry because previous write still in progress, it just takes long or may be hanging!
[INFO ] 2022-01-05 14:29:51.994 [[main]<azure_blob_storage] azureblobstorage - Skipped writing the registry because previous write still in progress, it just takes long or may be hanging!
[INFO ] 2022-01-05 14:29:52.402 [[main]<azure_blob_storage] azureblobstorage - Skipped writing the registry because previous write still in progress, it just takes long or may be hanging!
[INFO ] 2022-01-05 14:29:52.998 [[main]<azure_blob_storage] azureblobstorage - Skipped writing the registry because previous write still in progress, it just takes long or may be hanging!

However, it never got to write to data from the new blobs, because of the too small interval.
The offsets in the registry.dat don't contain info about the new blobs, as well.

We get the same [INFO ] log messages (with skipped writing...) in prod.

So, the issue might not be from the resume policy, but from the interval. The more data the storage account has, the higher the interval should be, to be able to write the data and the offsets to the registry. Otherwise, the processing will remain behind, like it happens now, when the interval is set to 60. Maybe 300 will do.

**Comment by laurentiubanica** [comment]
We still get these messages:
[2022-01-06T10:04:13,790][INFO ][logstash.inputs.azureblobstorage][][] Skipped writing the registry because previous write still in progress, it just takes long or may be hanging!
And also, I see streaming events from yesterday.

**Comment by janmg** [comment]
"Skipped writing the registry" is a notification when saving the registry is done in a separate thread. The variable busy_writing_registry is set to true to prevent multiple threads from writing into the registry, this would otherwise corrupt the registry. The better way is to use a mutex to do semaphore locking so that the code can only be executed by one thread at a time.

The original plugin logstash-input-azureblob uses blob file locking, but suffers from locks not being released, hanging the plugin.

There should only be one instance trying to write the registry at the same time. Having two logstash'es on the same or different machine processing the queue at the same time is a synchronization nightmare. It requires too much logic. The problem with ruby is that it uses a lot of global variables that are valid with the same JVM, this is a thread-safety problem which will not occur with only one instance running. 

The registry is filled at the startup by marshaling the the registry file. it is never read after that. Your registry.txt is corrupted. the content should be a has with the filename: {:lenth, :offset} to remember which file has been read to which offset in bytes. then the blobstore is listed in a filelist and a new registry is filled with the offsets of the loaded registry. With this new registry a worklist is selected for all files that are larger than it's offset. Those are processed and all the separate events within the file are send to the queue where logstash picks them up for further processing (filter and output). Once processing is done the plugin saves the registry if the processed events counter has increased since the last time. It looks then at the watch and then sleeps until the interval is over.

The registry can live on the local filepath of the logstash instance or in the remote blobstore. In 0.12.0 the azure-storage-blob dependency was updated to 2.0.3 or higher, this fixes a nokogiri/faraday vulnerability, but the newer version doesn't send a message on exceptions, the exception handling in my plugin then needs to account for that. Exceptions that happen for instance when archived blobs can't be read.

I can't put my finger on where the worklist decides to process the files again, there must be a variable that looses the current process tracking. This can happen due to thread safety issues where variables are overwritten, maybe in the filelist variable? This issue happens in the syslog output plugin where there is a "message" value, which is overlapping the variable message which reused, it is unfortunately also the name of the root of the logstash event

https://www.elastic.co/guide/en/logstash/current/plugins-outputs-syslog.html#plugins-outputs-syslog-message

**Comment by laurentiubanica** [comment]
I changed the registry location to local storage.
[main.txt](https://github.com/janmg/logstash-input-azure_blob_storage/files/7821894/main.txt)
Not sure if this one is ok, I do see offset and length. The behavior is the same. After it processes everything in the storage, it restarts.

**Comment by janmg** [comment]
Running this

@registry = Hash.new
@registry = Marshal.load(File.read("main.txt"))
puts @registry
puts
@registry.each do |name, file|
  puts name
  puts "offset: #{file[:offset]}"
  puts "length: #{file[:length]}"
end

Yields this:
resourceId=...d=03/h=13.../PT1H.json
offset: 134246
length: 134246
resourceId=...d=03/h=14.../PT1H.json
offset: 0
length: 62291

This means hour 13 is processed and shouldn't process anymore, hour 14 is not yet processed or at least the registry isn't saved with that information. That should happen after the processing interval or when the plugin is stopped.

Than the question is still, what the variable @registry contains and why the plugin creates a worklist that contains files that were already processed?

**Comment by laurentiubanica** [comment]
```
[INFO ] 2022-01-06 15:49:48.138 [[main]-pipeline-manager] javapipeline - Pipeline started {"pipeline.id"=>"main"}
[INFO ] 2022-01-06 15:49:48.233 [Agent thread] agent - Pipelines running {:count=>1, :running_pipelines=>[:main], :non_running_pipelines=>[]}
[INFO ] 2022-01-06 15:49:51.440 [[main]<azure_blob_storage] azureblobstorage - learn_encapsulation, this can be skipped by setting skip_learning => true. Or set both head_file and tail_file
[INFO ] 2022-01-06 15:49:51.980 [[main]<azure_blob_storage] azureblobstorage - learn json one of the attempts failed undefined method `name' for nil:NilClass
[INFO ] 2022-01-06 15:49:52.017 [[main]<azure_blob_storage] azureblobstorage - learn json one of the attempts failed undefined method `name' for nil:NilClass
[INFO ] 2022-01-06 15:49:52.050 [[main]<azure_blob_storage] azureblobstorage - learn json one of the attempts failed undefined method `name' for nil:NilClass
[INFO ] 2022-01-06 15:49:52.052 [[main]<azure_blob_storage] azureblobstorage - head will be: {"records":[ and tail is set to ]}
[INFO ] 2022-01-06 15:49:52.316 [[main]<azure_blob_storage] azureblobstorage - 1: list_blobs resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=13/m=00/macAddress=000D3ACCFD22/PT1H.json 0 134246
[INFO ] 2022-01-06 15:49:52.316 [[main]<azure_blob_storage] azureblobstorage - 1: list_blobs resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=14/m=00/macAddress=000D3ACCFD22/PT1H.json 0 62291
[INFO ] 2022-01-06 15:49:52.316 [[main]<azure_blob_storage] azureblobstorage - 1: list_blobs resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=18/m=00/macAddress=000D3ACCFD22/PT1H.json 0 104496
[INFO ] 2022-01-06 15:49:52.316 [[main]<azure_blob_storage] azureblobstorage - 1: list_blobs resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=19/m=00/macAddress=000D3ACCFD22/PT1H.json 0 130749
[INFO ] 2022-01-06 15:49:52.317 [[main]<azure_blob_storage] azureblobstorage - 1: list_blobs resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=20/m=00/macAddress=000D3ACCFD22/PT1H.json 0 131809
[INFO ] 2022-01-06 15:49:52.319 [[main]<azure_blob_storage] azureblobstorage - list_blobs took 0 sec
[INFO ] 2022-01-06 15:49:52.324 [[main]<azure_blob_storage] azureblobstorage - 2: adding offsets: resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=13/m=00/macAddress=000D3ACCFD22/PT1H.json 0 134246
[INFO ] 2022-01-06 15:49:52.325 [[main]<azure_blob_storage] azureblobstorage - 2: adding offsets: resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=14/m=00/macAddress=000D3ACCFD22/PT1H.json 0 62291
[INFO ] 2022-01-06 15:49:52.325 [[main]<azure_blob_storage] azureblobstorage - 2: adding offsets: resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=18/m=00/macAddress=000D3ACCFD22/PT1H.json 0 104496
[INFO ] 2022-01-06 15:49:52.326 [[main]<azure_blob_storage] azureblobstorage - 2: adding offsets: resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=19/m=00/macAddress=000D3ACCFD22/PT1H.json 0 130749
[INFO ] 2022-01-06 15:49:52.326 [[main]<azure_blob_storage] azureblobstorage - 2: adding offsets: resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=20/m=00/macAddress=000D3ACCFD22/PT1H.json 0 131809
[INFO ] 2022-01-06 15:49:52.328 [[main]<azure_blob_storage] azureblobstorage - worklist contains 5 blobs
[INFO ] 2022-01-06 15:49:52.331 [[main]<azure_blob_storage] azureblobstorage - 3: processing resourceId=/SUBSCRIPTIONS/11575A7E-D307-4203-96B4-0A03B5C3E912/RESOURCEGROUPS/IOTA-JAEA-RG-NETW/PROVIDERS/MICROSOFT.NETWORK/NETWORKSECURITYGROUPS/IOTA-JAEA-NSG-JUMPHOST/y=2022/m=01/d=03/h=13/m=00/macAddress=000D3ACCFD22/PT1H.json from 0 to 134246
[INFO ] 2022-01-06 15:50:04.704 [[main]<azure_blob_storage] azureblobstorage - going to sleep for 58 seconds
[INFO ] 2022-01-06 15:51:03.013 [[main]<azure_blob_storage] azureblobstorage - list_blobs took 1 sec
[INFO ] 2022-01-06 15:51:03.017 [[main]<azure_blob_storage] azureblobstorage - worklist contains 5 blobs
[INFO ] 2022-01-06 15:51:03.415 [Ruby-0-Thread-35: :1] azureblobstorage - processed 8082 events, saving 5 blobs and offsets to local registry /usr/share/logstash/plugin/main
[INFO ] 2022-01-06 15:51:07.768 [[main]<azure_blob_storage] azureblobstorage - going to sleep for 59 seconds
[INFO ] 2022-01-06 15:52:07.021 [[main]<azure_blob_storage] azureblobstorage - list_blobs took 1 sec
[INFO ] 2022-01-06 15:52:07.022 [[main]<azure_blob_storage] azureblobstorage - worklist contains 5 blobs
[INFO ] 2022-01-06 15:52:07.159 [Ruby-0-Thread-36: :1] azureblobstorage - processed 14612 events, saving 5 blobs and offsets to local registry /usr/share/logstash/plugin/main
[INFO ] 2022-01-06 15:52:09.381 [[main]<azure_blob_storage] azureblobstorage - going to sleep for 59 seconds
[INFO ] 2022-01-06 15:53:08.568 [[main]<azure_blob_storage] azureblobstorage - list_blobs took 0 sec
[INFO ] 2022-01-06 15:53:08.570 [[main]<azure_blob_storage] azureblobstorage - worklist contains 5 blobs
[INFO ] 2022-01-06 15:53:08.674 [Ruby-0-Thread-37: :1] azureblobstorage - processed 21142 events, saving 5 blobs and offsets to local registry /usr/share/logstash/plugin/main
[INFO ] 2022-01-06 15:53:10.647 [[main]<azure_blob_storage] azureblobstorage - going to sleep for 60 seconds
[INFO ] 2022-01-06 15:54:10.869 [[main]<azure_blob_storage] azureblobstorage - list_blobs took 0 sec
[INFO ] 2022-01-06 15:54:10.870 [[main]<azure_blob_storage] azureblobstorage - worklist contains 5 blobs
[INFO ] 2022-01-06 15:54:10.976 [Ruby-0-Thread-38: :1] azureblobstorage - processed 27672 events, saving 5 blobs and offsets to local registry /usr/share/logstash/plugin/main
[INFO ] 2022-01-06 15:54:13.640 [[main]<azure_blob_storage] azureblobstorage - going to sleep for 59 seconds
[INFO ] 2022-01-06 15:55:12.815 [[main]<azure_blob_storage] azureblobstorage - list_blobs took 0 sec
[INFO ] 2022-01-06 15:55:12.816 [[main]<azure_blob_storage] azureblobstorage - worklist contains 5 blobs
[INFO ] 2022-01-06 15:55:12.947 [Ruby-0-Thread-39: :1] azureblobstorage - processed 34202 events, saving 5 blobs and offsets to local registry /usr/share/logstash/plugin/main
[INFO ] 2022-01-06 15:55:15.492 [[main]<azure_blob_storage] azureblobstorage - going to sleep for 59 seconds
[INFO ] 2022-01-06 15:56:14.625 [[main]<azure_blob_storage] azureblobstorage - list_blobs took 0 sec
[INFO ] 2022-01-06 15:56:14.626 [[main]<azure_blob_storage] azureblobstorage - worklist contains 5 blobs
[INFO ] 2022-01-06 15:56:14.765 [Ruby-0-Thread-40: :1] azureblobstorage - processed 40732 events, saving 5 blobs and offsets to local registry /usr/share/logstash/plugin/main
[INFO ] 2022-01-06 15:56:16.791 [[main]<azure_blob_storage] azureblobstorage - going to sleep for 60 seconds
[INFO ] 2022-01-06 15:57:16.944 [[main]<azure_blob_storage] azureblobstorage - list_blobs took 0 sec
[INFO ] 2022-01-06 15:57:16.946 [[main]<azure_blob_storage] azureblobstorage - worklist contains 5 blobs
[INFO ] 2022-01-06 15:57:17.045 [Ruby-0-Thread-41: :1] azureblobstorage - processed 47262 events, saving 5 blobs and offsets to local registry /usr/share/logstash/plugin/main
[INFO ] 2022-01-06 15:57:19.794 [[main]<azure_blob_storage] azureblobstorage - going to sleep for 60 seconds
```

**Comment by laurentiubanica** [comment]
It has added all the offsets. However, is says it processed from 0 to 134246. After that, it list all the lobs, it doesn't say until which offset it has processed, but the number of the processed events is increasing from cycle to cycle (interval => 60).
However, it reaches the last filename (the one with h=20).

**Comment by laurentiubanica** [comment]
I found the issue.
There was a problem with @registry variable.

At line 216 from azure_blob_storage.rb:
@registry = newreg

I commented this line and now is working fine. I'll upload the logs.

**Comment by laurentiubanica** [comment]
[log.txt](https://github.com/janmg/logstash-input-azure_blob_storage/files/7829846/log.txt)
[main.txt](https://github.com/janmg/logstash-input-azure_blob_storage/files/7829848/main.txt)

I changed the extension of the registry file, to be able to upload id.

**Comment by janmg** [comment]
I pushed 0.12.3 ... I don't fully understand why replacing the @registry = newreg worked for 0.11.0 but stopped working for 0.12.0, but it has to do something with hashes and how they are cloned. Now I removed the newreg and replaced it with a loop that removes all the registry items that are not part of the listed files. So that files that have been removed are not kept forever on the registry. That was the reason to have a newreg in the first place.

I also replaced the variable in the registry_save for a proper mutex and removed the variable passing as @registry is a global var anyway.

**Comment by laurentiubanica** [comment]
Thank you for the new version !

**Comment by janmg** [comment]
Assume it's working fine now

**Comment by laurentiubanica** [comment]
Working just fine !
